src/scoring/hpf_scoring.py

```diff
@@ -159,25 +124,33 @@ def predict(self, input_stack):
             return [], {}, list(missing_packages)
         manifest_match = self.match_manifest(input_id_set)
         if manifest_match > 0:
-            companion_recommendation = self.recommend_known_user(manifest_match, input_id_set)
+            result, user_id = self.recommend_known_user(manifest_match, input_id_set)
+            companion_recommendation = self.filter_recommendation(result, input_id_set, user_id)
         else:
-            companion_recommendation = self.recommend_new_user(list(input_id_set))
+            result, user_id = self.recommend_new_user(list(input_id_set))
+            companion_recommendation = self.filter_recommendation(result, input_id_set, user_id)
         return companion_recommendation, package_topic_dict, list(missing_packages)
 
-    def match_feedback_manifest(self, input_id_set):
-        """Find a feedback manifest that matches user's input package list and return its index.
+
+    def match_manifest(self, input_id_set):  # pragma: no cover
+        """Find a manifest list that matches user's input package list and return its index.
 
         :param input_id_set: A set containing package ids of user's input package list.
-        :return manifest_id: The index of the matched feedback manifest.
+        :return manifest_id: The index of the matched manifest.
         """
-        for manifest_id, dependency_set in self.feedback_id_dict.items():
-            if dependency_set == input_id_set:
-                break
-        else:
-            manifest_id = -1
-        _logger.debug(
-                "input_id_set {} and feedback_manifest_id {}".format(input_id_set, manifest_id))
-        return int(manifest_id)
+        # TODO: Go back to the difference based logic, this simpler logic will do for now.
+        for manifest_id, dependency_set in self.manifest_id_dict.items():
+            if input_id_set.issubset(dependency_set):
+                current_app.logger.debug(
+                        "input_id_set {} and manifest_id {}".format(input_id_set, manifest_id))
+                return int(manifest_id)
+        return -1
+
+    def package_labelling(self, package_list):
+            package_data = self.package_id_dict.items()
+            labeled_packages = dict((v,k) for k,v in package_data)
```

Comment:
Isn't it better to create a reverse dict only once while the object is initialized and use it here, instead of re-creating it over and over again?